gridstatus/lmp_config.py

```diff
@@ -0,0 +1,211 @@
+import datetime
+import functools
+
+import pandas as pd
+
+from gridstatus.base import ISOBase, Markets, NotSupported
+
+
+class lmp_config:
+
+    configs = {}
+
+    def __init__(self, supports, tz=None):
+        self.supports = supports
+        self.tz = tz
+
+    def __call__(self, func):
+        lmp_config.configs[func.__qualname__] = self.supports
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if self._is_class_method(args, kwargs):
+                instance = args[0]
+                func_args = args[1:]
+                func_kwargs = kwargs
+                func_sig = self._parse_signature(
+                    instance,
+                    func_args,
+                    func_kwargs,
+                    tz=self.tz,
+                )
+                return self._class_method_wrapper(instance, func, func_sig)
+            else:
+                func_sig = self._parse_signature(None, args, kwargs, tz=self.tz)
+                return self._function_wrapper(func, func_sig)
+
+        return wrapper
+
+    @staticmethod
+    def _is_class_method(args, kwargs):
+        # args[0] must be "self" and an instance of ISOBase
+        return len(args) > 0 and isinstance(args[0], ISOBase)
+
+    @staticmethod
+    def _validate_market_arg_signature(market):
+        return isinstance(market, str) or isinstance(market, Markets)
+
+    @staticmethod
+    def _parse_date(date, tz):
+        if date == "latest":
+            return lmp_config.__handle_date("today", tz=tz)
+        elif (
+            isinstance(date, str)
+            or isinstance(date, pd.Timestamp)
+            or isinstance(date, datetime.date)
+        ):
+            final_date = lmp_config.__handle_date(date, tz=tz)
+            if not isinstance(final_date, pd.Timestamp):
+                raise ValueError(f"Cannot parse date {repr(date)}")
+            return final_date
+        else:
+            raise ValueError(
+                "date must be string or pd.Timestamp: "
+                f"{repr(date)} of type ({type(date)})",
+            )
+
+    @staticmethod
+    def _class_method_wrapper(instance, func, func_sig):
+        instance_args = tuple([instance] + list(func_sig["args"]))
+        instance_kwargs = func_sig["kwargs"]
+        return func(*instance_args, **instance_kwargs)
+
+    @staticmethod
+    def _function_wrapper(func, func_sig):
+        return func(*func_sig["args"], **func_sig["kwargs"])
+
+    @staticmethod
+    def _parse_market(market):
+        return Markets(market)
+
+    def _parse_signature(self, instance, args, kwargs, tz):
```

Comment:
I'm not sure how to simplify right now, but adding all this higher-level logic to parsing the signature seems potentially less maintainable and eventually more error prone than doing nothing right now. 

one approach might be to put some of this logic for standardizing date/start/end in a different decorator. there is some similar code doing this validation for support date range here: https://github.com/kmax12/gridstatus/blob/860565ae81ae888df5e3442bfe9c12a2cbac45f7/gridstatus/decorators.py#L38

Reply:
Yeah this whole method is... complex. I only saw the decorator stuff after putting up this PR - so I'll see how we can use args_dict here too.

Reply:
OK I cleaned this up quite a bit: similar to `_get_args_dict`, I have `_get_fn_params` return OrderedDict `args`, and `kwargs` as-is. That allows me to recompose the original args after validating and transforming (if needed) `date` and `market`. The replacement logic is similarly simplified.